src/parquet/encodings/plain-encoding.h

```diff
@@ -163,19 +163,41 @@ class PlainEncoder : public Encoder<DType> {
 
   explicit PlainEncoder(
       const ColumnDescriptor* descr, MemoryAllocator* allocator = default_allocator())
-      : Encoder<DType>(descr, Encoding::PLAIN, allocator) {}
+      : Encoder<DType>(descr, Encoding::PLAIN, allocator),
+        values_sink_(new InMemoryOutputStream(IN_MEMORY_DEFAULT_CAPACITY, allocator)) {}
 
-  void Encode(const T* src, int num_values, OutputStream* dst) override;
+  std::shared_ptr<Buffer> FlushValues() override;
+  void Put(const T* src, int num_values) override;
+  void Encode(const T* src, int num_values, OutputStream* dst);
+
+ protected:
+  std::shared_ptr<InMemoryOutputStream> values_sink_;
```

Comment:
Per the above discussion about accumulating data pages in memory, we may consider nixing OutputStream from the encoder API. May need to do some more analysis of how other implementations have had to deal with this. For example, parquet-mr has

https://github.com/apache/parquet-mr/blob/master/parquet-hadoop/src/main/java/org/apache/parquet/hadoop/ColumnChunkPageWriteStore.java#L81

and then the concatenated data pages are written out to the output stream after the dictionary page (if there is one)

https://github.com/apache/parquet-mr/blob/master/parquet-hadoop/src/main/java/org/apache/parquet/hadoop/ColumnChunkPageWriteStore.java#L187